File: pending/biglist/src/main/java/onl/ycode/tmaker/Node.java
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis

package onl.ycode.tmaker;

import onl.ycode.stormify.FieldInfo;
import onl.ycode.stormify.TableInfo;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static java.util.Collections.emptyMap;
import static onl.ycode.stormify.StormifyManager.stormify;
import static onl.ycode.tmaker.DefaultDataConverter.validate;

class Node {
    private final String tableAlias;
    private final Node parent;
    private final TableInfo tableInfo;
    private final FieldInfo fieldInfo;
    private final Map<String, Node> children;
    private boolean active = false;

    Node(String tableAlias, Node parent, TableInfo tableInfo, FieldInfo fieldInfo) {
        this.tableAlias = tableAlias;
        this.parent = parent;
        this.tableInfo = tableInfo;
        this.fieldInfo = fieldInfo;
        this.children = tableInfo == null ? emptyMap() : new LinkedHashMap<>();
    }

    Node findChild(String fieldName, AtomicInteger tableCounter) {
        if (isLeaf())
            throw new IllegalStateException("Field " + fieldInfo.getName() + " is not a reference field");
        validate(fieldName);
        Node child = children.get(fieldName);
        if (child == null) {
            FieldInfo childField = tableInfo.getField(fieldName);
            if (childField == null)
                throw new IllegalArgumentException("Field " + fieldName + " not found in table " + tableInfo.getClassType().getName());
            TableInfo childTable;
            String tableName;
            if (childField.isReference()) {
                childTable = stormify().getTableInfo(childField.getType());
                tableName = "t" + tableCounter.incrementAndGet();
            } else {
                childTable = null;
                tableName = null;
            }
            child = new Node(tableName, this, childTable, childField);
            children.put(fieldName, child);
        }
        return child;
    }

    boolean isLeaf() {
        return tableInfo == null;
    }

    // BotToM-up activation
    void activate() {
        active = true;
        if (parent != null)
            parent.activate();
    }

    // Top-down deactivation
    void deactivate() {
        active = false;
        for (Node child : children.values())
            child.deactivate();
    }

    void getForeignKeys(StringBuilder out) {
        if (isLeaf())
            return;
        for (Node child : children.values()) {
            if (!child.active || child.isLeaf())
                continue;
            out
                    .append(" JOIN ")
                    .append(child.tableInfo.getTableName())
                    .append(" ")
                    .append(child.tableAlias)
                    .append(" ON ")
                    .append(child.tableAlias)
                    .append(".")
                    .append(child.tableInfo.getPrimaryKey().getDbName())
                    .append(" = ")
                    .append(tableAlias.isEmpty() ? tableInfo.getTableName() : tableAlias)
                    .append(".")
                    .append(child.fieldInfo.getDbName());
        }
        for (Node child : children.values())
            child.getForeignKeys(out);
    }

    String getTableAlias() {
        return isLeaf() ? parent.tableAlias : tableAlias;
    }

    String getTableName() {
        String alias = getTableAlias();
        if (!alias.isEmpty())
            return alias;
        return isLeaf() ? parent.tableInfo.getTableName() : tableInfo.getTableName();
    }

    String getColumnName() {
        if (!isLeaf())
            throw new IllegalStateException("Node is not a leaf");
        return getTableName() + "." + fieldInfo.getDbName();
    }

    Class<?> getType() {
        return fieldInfo.getType();
    }
}

File: db/src/commonMain/kotlin/onl/ycode/stormify/biglist/Node.kt
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis
package onl.ycode.stormify.biglist

import onl.ycode.stormify.TableInfo
import onl.ycode.stormify.isScalarClass
import kotlin.reflect.KClass


internal sealed class Node(val type: KClass<*>, internal val invalidate: () -> Unit, private val parent: Node?) {
    abstract val children: Map<String, Node>
    abstract fun findChild(fieldName: String, tableCounter: () -> Int): Node
    abstract fun getForeignKeys(out: StringBuilder)

    var active = false

    // Bottom-up activation
    fun activate() {
        active = true
        parent?.activate()
    }

    // Top-down deactivation
    fun deactivate() {
        active = false
        for (child in children.values)
            child.deactivate()
    }

}

internal class NodeField(
    internal val columnHandler: String,
    type: KClass<*>,
    invalidate: () -> Unit,
    parent: Node?
) : Node(type, invalidate, parent) {
    override val children = emptyMap<String, Node>()
    override fun findChild(fieldName: String, tableCounter: () -> Int) =
        throw UnsupportedOperationException("Field node does not have children")

    override fun getForeignKeys(out: StringBuilder) = Unit
}

internal class NodeTable(
    tableAlias: String,
    private val parentReference: String,
    type: KClass<*>,
    invalidate: () -> Unit,
    parent: Node?
) :
    Node(type, invalidate, parent) {
    override val children = mutableMapOf<String, Node>()
    private val tableInfo = TableInfo.retrieve(type)

    private val tableDefinition = if (tableAlias.isEmpty()) tableInfo.table else "${tableInfo.table} AS $tableAlias"
    private val tableHandler = tableAlias.ifEmpty { tableInfo.table }

    override fun findChild(fieldName: String, tableCounter: () -> Int) = children[fieldName] ?: run {
        val childType = tableInfo.getType(fieldName)
        val node = if (isScalarClass(childType))
            NodeField("$tableHandler.$fieldName", childType, invalidate, this)
        else
            NodeTable("t${tableCounter()}", fieldName, childType, invalidate, this)
        children[fieldName] = node
        return node
    }

    override fun getForeignKeys(out: StringBuilder) {
        for (child in children.values) {
            if (!child.active || child !is NodeTable) continue
            out
                .append(" JOIN ")
                .append(child.tableDefinition)
                .append(" ON ")
                .append(tableHandler)
                .append(".")
                .append(child.tableInfo.table)
                .append(" = ")
                .append(child.tableHandler)
                .append(".")
                .append(child.parentReference)
        }
        for (child in children.values)
            child.getForeignKeys(out)
    }
}
